### PR TITLE
[minmax-embedder-value-test] Introduce tests for minmax-embedder

### PR DESCRIPTION
**Pull request by glistening**
It introduces tests for minmax-embedder.

ONE-DCO-1.0-Signed-off-by: Sanggyu Lee <sg5.lee@samsung.com>

Related: #11668

### CODE REVIEW

**Comment by glistening** [comment]
### Generate minmax-data

`minmax-data-gen` generates (min, max) sequence by simple increment.
```
$ build/compiler/minmax-embedder-test/gen/minmax-data-gen mm.h5
$ h5dump mm.h5
HDF5 "mm.h5" {
GROUP "/" {
   GROUP "value" {
      GROUP "run_0" {
         GROUP "model_0" {
            GROUP "subg_0" {
               DATASET "input_0" {
                  DATATYPE  H5T_IEEE_F32BE
                  DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
                  DATA {
                  (0): 0, 1
                  }
               }
               DATASET "op_0" {
                  DATATYPE  H5T_IEEE_F32BE
                  DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
                  DATA {
                  (0): 2, 3
                  }
...
```
Run once more, it appends 1 run.
`minmax-data-gen` increase the starting number by 1000 for each new run.
```
$ build/compiler/minmax-embedder-test/gen/minmax-data-gen mm.h5
$ h5dump mm.h5
HDF5 "mm.h5" {
GROUP "/" {
   GROUP "value" {
      GROUP "run_0" {
         GROUP "model_0" {
            GROUP "subg_0" {
               DATASET "input_0" {
                  DATATYPE  H5T_IEEE_F32BE
                  DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
                  DATA {
                  (0): 0, 1
                  }
               }
               DATASET "op_0" {
                  DATATYPE  H5T_IEEE_F32BE
                  DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
                  DATA {
                  (0): 2, 3
                  }
               }
            }
         }
      }
      GROUP "run_1" {
         GROUP "model_0" {
            GROUP "subg_0" {
               DATASET "input_0" {
                  DATATYPE  H5T_IEEE_F32BE
                  DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
                  DATA {
                  (0): 1000, 1001
                  }
               }
               DATASET "op_0" {
                  DATATYPE  H5T_IEEE_F32BE
                  DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
                  DATA {
                  (0): 1002, 1003
                  }
...
```

### Embed minmax data

Let's confirm there is no (min, max) before embedding (min, max).

<sub> `Abs_000.circle` comes from `compiler/common-artifacts`. </sub>
```
$ build/compiler/circledump/circledump Abs_000.circle 
Dump: Abs_000.circle

...
T(0:0) FLOAT32 (1, 3, 3, 2) B(1) ifm

T(0:1) FLOAT32 (1, 3, 3, 2) B(2) ofm
```
Then, embed the minmax-data generated above using `minmax-embedder`
```
$ build/compiler/minmax-embedder/minmax-embedder --min_percentile 0 --max_percentile 100 Abs_000.circle mm.h5 -o out2.circle

$ build/compiler/circledump/circledump out2.circle
Dump: out2.circle
...
Operands: T(subgraph index : tensor index) TYPE (shape) (shape_signature) B(buffer index) (variable) OperandName
T(0:0) FLOAT32 (1, 3, 3, 2) B(1) ifm
    Quantization: min(0) max(1001) quantized_dimension(0)

T(0:1) FLOAT32 (1, 3, 3, 2) B(2) ofm
    Quantization: min(2) max(1003) quantized_dimension(0)
```
You can see `min` and `max` for `ifm` (= input_0) and `ofm` (= op_0's output).

**Comment by glistening** [comment]
## Two more cases

### Add_000 (input: 2, op: 1, output: 1)
```sh
# 3 runs, 21 means 2 inputs and 1 op
$ build/debug/compiler/minmax-embedder-test/gen/minmax-data-gen --num_inputs 2 21.h5
$ build/debug/compiler/minmax-embedder-test/gen/minmax-data-gen --num_inputs 2 21.h5
$ build/debug/compiler/minmax-embedder-test/gen/minmax-data-gen --num_inputs 2 21.h5

$ h5dump 21.h5 
HDF5 "2in1op.h5" {
GROUP "/" {
   GROUP "value" {
      GROUP "run_0" {
         GROUP "model_0" {
            GROUP "subg_0" {
               DATASET "input_0" {
                  DATA {
                  (0): 0, 1
               }
               DATASET "input_1" {
                  DATA {
                  (0): 2, 3
                  }
               }
               DATASET "op_0" {
                  DATA {
                  (0): 4, 5
                  }
               }
            }
         }
      }
      GROUP "run_1" {
         GROUP "model_0" {
            GROUP "subg_0" {
               DATASET "input_0" {
                  DATA {
                  (0): 1000, 1001
                  }
               }
               DATASET "input_1" {
                  DATA {
                  (0): 1002, 1003
                  }
               }
               DATASET "op_0" {
                  DATA {
                  (0): 1004, 1005
                  }
               }
            }
         }
      }
      GROUP "run_2" {
         GROUP "model_0" {
            GROUP "subg_0" {
               DATASET "input_0" {
                  DATA {
                  (0): 2000, 2001
                  }
               }
               DATASET "input_1" {
                  DATA {
                  (0): 2002, 2003
                  }
               }
               DATASET "op_0" {
                  DATA {
                  (0): 2004, 2005
                  }
               }
...
```
```
$ build/debug/compiler/minmax-embedder/minmax_embedder \
--min_percentile 0 --max_percentile 100 \
-o Add_000.circle{.out,} 21.h5

$ build/debug/compiler/circledump/circledump Add_000.circle.out 
...
T(0:0) FLOAT32 (1, 4, 4, 3) B(1) ifm1
    Quantization: min(0) max(2001) quantized_dimension(0)

T(0:1) FLOAT32 (1, 4, 4, 3) B(2) ifm2
    Quantization: min(2) max(2003) quantized_dimension(0)

T(0:2) FLOAT32 (1, 4, 4, 3) B(3) ofm
    Quantization: min(4) max(2005) quantized_dimension(0)

Operators: O(subgraph index : operator index) OpCodeName 
    Option(values) ... <-- depending on OpCode
    I T(tensor index) OperandName <-- as input
    O T(tensor index) OperandName <-- as output
O(0:0) ADD 
    Activation(NONE) 
    I T(0:0) ifm1
    I T(0:1) ifm2
    O T(0:2) ofm

Inputs/Outputs: I(input)/O(output) T(tensor index) OperandName
I T(0:0) ifm1
I T(0:1) ifm2
O T(0:2) ofm
```

### Net_Conv_Relu6_000 (input: 1, op: 4, output: 1)
```
# 2 runs
$ build/debug/compiler/minmax-embedder-test/gen/minmax-data-gen --num_ops 4 14.h5
$ build/debug/compiler/minmax-embedder-test/gen/minmax-data-gen --num_ops 4 14.h5
$ h5dump 14.h5
HDF5 "14.h5" {
GROUP "/" {
   GROUP "value" {
      GROUP "run_0" {
         GROUP "model_0" {
            GROUP "subg_0" {
               DATASET "input_0" {
                  DATA {
                  (0): 0, 1
                  }
               }
               DATASET "op_0" {
                  DATA {
                  (0): 2, 3
                  }
               }
               DATASET "op_1" {
                  DATA {
                  (0): 4, 5
                  }
               }
               DATASET "op_2" {
                  DATA {
                  (0): 6, 7
                  }
               }
               DATASET "op_3" {
                  DATA {
                  (0): 8, 9
                  }
               }
            }
         }
      }
      GROUP "run_1" {
         GROUP "model_0" {
            GROUP "subg_0" {
               DATASET "input_0" {
                  DATA {
                  (0): 1000, 1001
                  }
               }
               DATASET "op_0" {
                  DATA {
                  (0): 1002, 1003
                  }
               }
               DATASET "op_1" {
                  DATA {
                  (0): 1004, 1005
                  }
               }
               DATASET "op_2" {
                  DATA {
                  (0): 1006, 1007
                  }
               }
               DATASET "op_3" {
                  DATA {
                  (0): 1008, 1009
```
```
$ build/debug/compiler/minmax-embedder/minmax_embedder \
--min_percentile 0 --max_percentile 100 \
-o Net_Conv_Relu6_000.circle{.out,} 14.h5

$ build/debug/compiler/circledump/circledump Net_Conv_Relu6_000.circle.out
T(0:0) FLOAT32 (1, 16, 16, 3) B(1) Placeholder
    Quantization: min(0) max(1001) quantized_dimension(0)

T(0:1) FLOAT32 (3, 3, 3, 3) B(2) Conv2D_1

T(0:2) FLOAT32 (3) B(3) Conv2D_2

T(0:3) FLOAT32 (1, 16, 16, 3) B(4) Conv2D_11
    Quantization: min(2) max(1003) quantized_dimension(0)

T(0:4) FLOAT32 (1, 16, 16, 3) B(5) ReLU6
    Quantization: min(4) max(1005) quantized_dimension(0)

T(0:5) FLOAT32 (3, 3, 3, 3) B(6) Conv2D_21

T(0:6) FLOAT32 (1, 16, 16, 3) B(7) Conv2D_22
    Quantization: min(6) max(1007) quantized_dimension(0)

T(0:7) FLOAT32 (1, 16, 16, 3) B(8) ReLU6_1
    Quantization: min(8) max(1009) quantized_dimension(0)

Operators: O(subgraph index : operator index) OpCodeName 
    Option(values) ... <-- depending on OpCode
    I T(tensor index) OperandName <-- as input
    O T(tensor index) OperandName <-- as output
O(0:0) CONV_2D 
    Padding(0) Stride.W(1) Stride.H(1) Dilation.W(1) Dilation.H(1) Activation(NONE)
    I T(0:0) Placeholder
    I T(0:1) Conv2D_1
    I T(0:2) Conv2D_2
    O T(0:3) Conv2D_11
O(0:1) RELU6 
    I T(0:3) Conv2D_11
    O T(0:4) ReLU6
O(0:2) CONV_2D 
    Padding(0) Stride.W(1) Stride.H(1) Dilation.W(1) Dilation.H(1) Activation(NONE)
    I T(0:4) ReLU6
    I T(0:5) Conv2D_21
    I T(0:2) Conv2D_2
    O T(0:6) Conv2D_22
O(0:3) RELU6 
    I T(0:6) Conv2D_22
    O T(0:7) ReLU6_1

Inputs/Outputs: I(input)/O(output) T(tensor index) OperandName
I T(0:0) Placeholder
O T(0:7) ReLU6_1
```

**Comment by glistening** [comment]
@jinevening, @hseok-oh Finally, it is ready for review.
```
$ ctest --test-dir build -R minmax_embedder_value* -V
Internal ctest changing into directory: /ssd/z/ONE/build
UpdateCTestConfiguration  from :/ssd/z/ONE/build/DartConfiguration.tcl
Parse Config file:/ssd/z/ONE/build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/ssd/z/ONE/build/DartConfiguration.tcl
Parse Config file:/ssd/z/ONE/build/DartConfiguration.tcl
Test project /ssd/z/ONE/build
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 109
    Start 109: minmax_embedder_value_test

109: Test command: /ssd/z/ONE/compiler/minmax-embedder-value-test/testall.sh "/ssd/z/ONE/build/compiler/minmax-embedder-value-test/test.config" "Abs_000" "Add_000" "Net_Conv_Relu6_000"
109: Test timeout computed to be: 1500
109: -- Found ARTIFACTS_PATH: /ssd/z/ONE/build/compiler/common-artifacts
109: -- Found MINMAX_DATA_GEN: /ssd/z/ONE/build/compiler/minmax-embedder-value-test/gen/minmax_data_gen
109: -- Found MINMAX_EMBEDDER: /ssd/z/ONE/build/compiler/minmax-embedder/minmax_embedder
109: -- Found CIRCLEDUMP: /ssd/z/ONE/build/compiler/circledump/circledump
109: -- Found CONFIG_PATH: /ssd/z/ONE/build/compiler/minmax-embedder-value-test/test.config
109: /ssd/z/ONE/build/compiler/minmax-embedder-value-test /ssd/z/ONE/build/compiler/minmax-embedder-value-test
109: /ssd/z/ONE/build/compiler/minmax-embedder-value-test
109: PASSED
1/1 Test #109: minmax_embedder_value_test .......   Passed    0.15 sec

The following tests passed:
	minmax_embedder_value_test

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   0.15 sec
```

For `Add_000.log`, You can see the full log for details.

```
cat /ssd/z/ONE/build/compiler/minmax-embedder-value-test/Add_000.log
```

Here is some of log.

```
declare -a INPUT_NAMES=([0]="ifm1" [1]="ifm2")
declare -a OP_OUT_NAMES=([0]="ofm")
declare -a MD_MAX=([0]="10007" [1]="10017")
declare -a MD_MIN=([0]="0" [1]="10")
declare -a OP_MIN=([0]="0")
declare -a OP_MAX=([0]="10007")
++ for i in "${!MD_MIN[@]}"
++ EXPECTED_MIN=0
++ EXPECTED_MAX=10007
++ [[ 0 != \0 ]]
++ [[ 10007 != \1\0\0\0\7 ]]
++ for i in "${!MD_MIN[@]}"
++ EXPECTED_MIN=10
++ EXPECTED_MAX=10017
++ [[ 10 != \1\0 ]]
++ [[ 10017 != \1\0\0\1\7 ]]
++ for i in "${!OP_MIN[@]}"
++ EXPECTED_MIN=0
++ EXPECTED_MAX=10007
++ [[ 0 != \0 ]]
++ [[ 10007 != \1\0\0\0\7 ]]
++ touch /ssd/z/ONE/build/compiler/minmax-embedder-value-test/Add_000.passed
```

**Comment by glistening** [comment]
## MinMax Data in HDF5

### Abs_000.minmax 
```
$ h5dump /ssd/z/ONE/build/compiler/minmax-embedder-value-test/Abs_000.minmax 
HDF5 "/ssd/z/ONE/build/compiler/minmax-embedder-value-test/Abs_000.minmax" {
GROUP "/" {
   GROUP "value" {
      GROUP "run_0" {
         GROUP "model_0" {
            GROUP "subg_0" {
               DATASET "input_0" {
                  DATATYPE  H5T_IEEE_F32BE
                  DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
                  DATA {
                  (0): 0, 7
                  }
               }
               DATASET "op_0" {
                  DATATYPE  H5T_IEEE_F32BE
                  DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
                  DATA {
                  (0): 0, 7
                  }
               }
            }
         }
      }
      GROUP "run_1" {
         GROUP "model_0" {
            GROUP "subg_0" {
               DATASET "input_0" {
                  DATATYPE  H5T_IEEE_F32BE
                  DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
                  DATA {
                  (0): 10000, 10007
                  }
               }
               DATASET "op_0" {
                  DATATYPE  H5T_IEEE_F32BE
                  DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
                  DATA {
                  (0): 10000, 10007
                  }
               }
            }
         }
      }
   }
}
}
```

### Add_000
```
$ h5dump /ssd/z/ONE/build/compiler/minmax-embedder-value-test/Add_000.minmax 
HDF5 "/ssd/z/ONE/build/compiler/minmax-embedder-value-test/Add_000.minmax" {
GROUP "/" {
   GROUP "value" {
      GROUP "run_0" {
         GROUP "model_0" {
            GROUP "subg_0" {
               DATASET "input_0" {
                  DATATYPE  H5T_IEEE_F32BE
                  DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
                  DATA {
                  (0): 0, 7
                  }
               }
               DATASET "input_1" {
                  DATATYPE  H5T_IEEE_F32BE
                  DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
                  DATA {
                  (0): 10, 17
                  }
               }
               DATASET "op_0" {
                  DATATYPE  H5T_IEEE_F32BE
                  DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
                  DATA {
                  (0): 0, 7
                  }
               }
            }
         }
      }
      GROUP "run_1" {
         GROUP "model_0" {
            GROUP "subg_0" {
               DATASET "input_0" {
                  DATATYPE  H5T_IEEE_F32BE
                  DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
                  DATA {
                  (0): 10000, 10007
                  }
               }
               DATASET "input_1" {
                  DATATYPE  H5T_IEEE_F32BE
                  DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
                  DATA {
                  (0): 10010, 10017
                  }
               }
               DATASET "op_0" {
                  DATATYPE  H5T_IEEE_F32BE
                  DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
                  DATA {
                  (0): 10000, 10007
                  }
               }
            }
         }
      }
   }
}
}
```

### Net_Conv_Relu6_000.minmax
```
$ h5dump /ssd/z/ONE/build/compiler/minmax-embedder-value-test/Net_Conv_Relu6_000.minmax
HDF5 "/ssd/z/ONE/build/compiler/minmax-embedder-value-test/Net_Conv_Relu6_000.minmax" {
GROUP "/" {
   GROUP "value" {
      GROUP "run_0" {
         GROUP "model_0" {
            GROUP "subg_0" {
               DATASET "input_0" {
                  DATATYPE  H5T_IEEE_F32BE
                  DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
                  DATA {
                  (0): 0, 7
                  }
               }
               DATASET "op_0" {
                  DATATYPE  H5T_IEEE_F32BE
                  DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
                  DATA {
                  (0): 0, 7
                  }
               }
               DATASET "op_1" {
                  DATATYPE  H5T_IEEE_F32BE
                  DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
                  DATA {
                  (0): 10, 17
                  }
               }
               DATASET "op_2" {
                  DATATYPE  H5T_IEEE_F32BE
                  DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
                  DATA {
                  (0): 20, 27
                  }
               }
               DATASET "op_3" {
                  DATATYPE  H5T_IEEE_F32BE
                  DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
                  DATA {
                  (0): 30, 37
                  }
               }
            }
         }
      }
      GROUP "run_1" {
         GROUP "model_0" {
            GROUP "subg_0" {
               DATASET "input_0" {
                  DATATYPE  H5T_IEEE_F32BE
                  DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
                  DATA {
                  (0): 10000, 10007
                  }
               }
               DATASET "op_0" {
                  DATATYPE  H5T_IEEE_F32BE
                  DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
                  DATA {
                  (0): 10000, 10007
                  }
               }
               DATASET "op_1" {
                  DATATYPE  H5T_IEEE_F32BE
                  DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
                  DATA {
                  (0): 10010, 10017
                  }
               }
               DATASET "op_2" {
                  DATATYPE  H5T_IEEE_F32BE
                  DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
                  DATA {
                  (0): 10020, 10027
                  }
               }
               DATASET "op_3" {
                  DATATYPE  H5T_IEEE_F32BE
                  DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
                  DATA {
                  (0): 10030, 10037
                  }
               }
            }
         }
      }
   }
}
}
```

**Comment by hseok-oh** [comment]
```
[2023-10-19T06:49:52.598Z] /opt/jenkins_agent/workspace/nnfw/master/pr-nncc-jammy-debug/build/compiler/minmax-embedder-value-test
[2023-10-19T06:49:52.598Z] FAILED
[2023-10-19T06:49:52.598Z] - Net_Conv_Relu6_000
[2023-10-19T06:49:52.598Z] \
[2023-10-19T06:49:52.598Z]         Start 116: luci_value_test
```

**Comment by glistening** [comment]
I confirmed `testall.sh` works after adding `-a` switch for `grep`.
`grep` considers the output of `circledump` as binary input.
Maybe for the part of dumping buffer values, because other 2 tests (`Add` and `Abs` are okay, they don't have constant buffer).

It does not happen on my Ubuntu, perhaps I set `LC_COLLATE=C`.
It happens on vanila Ubuntu (which does not set environment variables for locale).

File: compiler/minmax-embedder-value-test/gen/src/H5Writer.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    return parent.createGroup(child.c_str());
+  }
+}
+
+uint32_t to_u32(uint64_t v)
+{
+  if (v > UINT32_MAX)
+    throw std::overflow_error("to_u32 gets a value bigger than uint32 max.");
+  return static_cast<uint32_t>(v);
+}
```
Comment (jinevening) [inline compiler/minmax-embedder-value-test/gen/src/H5Writer.cpp:47]: Seems duplicate with the one in Cast.h

File: compiler/minmax-embedder-value-test/gen/src/H5Writer.h [hunk 2]
```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2019 Samsung Electronics Co., Ltd. All Rights Reserved
```
Comment (jinevening) [inline compiler/minmax-embedder-value-test/gen/src/H5Writer.h:2]:
```suggestion
 * Copyright (c) 2023 Samsung Electronics Co., Ltd. All Rights Reserved
```

File: compiler/minmax-embedder-value-test/gen/src/H5Writer.cpp [hunk 3]
```diff
@@ -39,13 +39,6 @@ H5::Group ensureGroup(H5::Group parent, const std::string &child)
   }
 }
 
-uint32_t to_u32(uint64_t v)
-{
-  if (v > UINT32_MAX)
-    throw std::overflow_error("to_u32 gets a value bigger than uint32 max.");
-  return static_cast<uint32_t>(v);
-}
-
```
Comment (hseok-oh) [inline compiler/minmax-embedder-value-test/gen/src/H5Writer.cpp]: Removed?

**Review by hseok-oh — APPROVED** [review]
LGTM

**Review by jinevening — APPROVED** [review]
LGTM

**Comment by glistening** [comment]
I took a closer look into https://github.com/Samsung/ONE/pull/11699#issuecomment-1771872647. The problem is not in Buffer printing, but in `Padding` printing. I will make a PR.

**Pull request merged**